chain/epoch_manager/src/lib.rs

```diff
@@ -397,35 +402,46 @@ impl EpochManager {
                     }
                 }
 
-                let BlockInfo { block_tracker, mut all_proposals, shard_tracker, .. } =
-                    prev_block_info;
-
-                // Update block produced/expected tracker.
-                block_info.update_block_tracker(
-                    &epoch_info,
-                    prev_block_info.height,
-                    if is_epoch_start { HashMap::default() } else { block_tracker },
-                );
-                block_info.update_shard_tracker(
-                    &epoch_info,
-                    prev_block_info.height,
-                    if is_epoch_start { HashMap::default() } else { shard_tracker },
-                );
-                // accumulate values
                 if is_epoch_start {
-                    block_info.all_proposals = block_info.proposals.clone();
                     self.save_epoch_start(
                         &mut store_update,
                         &block_info.epoch_id,
                         block_info.height,
                     )?;
-                } else {
-                    all_proposals.extend(block_info.proposals.clone());
-                    block_info.all_proposals = all_proposals;
                 }
 
                 // Save current block info.
                 self.save_block_info(&mut store_update, current_hash, block_info.clone())?;
+                let last_block_hash = match self.get_block_info(&block_info.last_final_block_hash) {
+                    Ok(final_block_info) => {
+                        if final_block_info.epoch_id != block_info.epoch_id {
+                            if is_epoch_start {
+                                Some(current_hash)
```

Comment:
Let me understand this logic:

1. We need that for all the future calls to the `get_and_update_epoch_info_aggregator` the cached block corresponds to one of the blocks in the ancestry of the `last_block`
2. Thus, it is safe to update the cached block up to the last final block.
3. If there's an epoch switch, and the current block is the first block in the epoch, it is safe to pass the block that is not final, because on the epoch switch `get_and_update_epoch_info_aggregator` will always recreate the aggregator, so even if there's a fork, we won't try to reuse the cache.

I see some problems with it:

1. A final block can be reverted. We should handle such a case gracefully. Seems like right now the aggregator  in case of a fork will merge all the proposals and tracking data from both forks (it will start with the data from the old fork, and will be merging backwards until it sees the last hash, or until the epoch start, but because it will never see the last hash, it will just merge in all the proposals)
2. The fork of a non-final block at the epoch start actually will result in an incorrect behavior. Consider

```

B <- A
  <-    A'
```

B is the last block of the prev epoch, A and A' are the first blocks of the new epoch. A is recognized first, and because `is_epoch_start` here is true, the cache is updated to be all the way up to `A`.
When A' arrives, here we again see `is_epoch_start`, and again set the `last_block_hash = current_hash`, and attempt to update the aggregator. The aggregator however will not consider it as `is_epoch_start`, because from the aggregator's perspective the last cached block and the new block are in the same epoch, so it will just merge their proposals and trackers, again resulting in a wrong stats.

Reply:
Fixed